--- a/libde265/CMakeLists.txt
+++ b/libde265/CMakeLists.txt
@@ -83,8 +83,6 @@ if(MSVC)
   )
 endif()
 
-add_definitions(-DLIBDE265_EXPORTS)
-
 add_subdirectory (encoder)
 
 if(SUPPORTS_SSE4_1)
@@ -92,8 +90,14 @@ if(SUPPORTS_SSE4_1)
   add_subdirectory (x86)
 endif()
 
-add_library(${PROJECT_NAME} SHARED ${libde265_sources} ${ENCODER_OBJECTS} ${X86_OBJECTS})
+add_library(${PROJECT_NAME} ${libde265_sources} ${ENCODER_OBJECTS} ${X86_OBJECTS})
 target_link_libraries(${PROJECT_NAME} PRIVATE Threads::Threads)
+set_target_properties(${PROJECT_NAME} PROPERTIES CXX_VISIBILITY_PRESET hidden VISIBILITY_INLINES_HIDDEN ON)
+if(BUILD_SHARED_LIBS)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE LIBDE265_EXPORTS HAVE_VISIBILITY=1)
+else()
+  target_compile_definitions(${PROJECT_NAME} PUBLIC LIBDE265_STATIC_BUILD)
+endif()
 
 write_basic_package_version_file(${PROJECT_NAME}ConfigVersion.cmake COMPATIBILITY ExactVersion)
 
--- a/libde265/de265.h
+++ b/libde265/de265.h
@@ -36,17 +36,17 @@ extern "C" {
 #endif
 #include <stdint.h>
 
-#if defined(_MSC_VER) && !defined(LIBDE265_STATIC_BUILD)
-  #ifdef LIBDE265_EXPORTS
-  #define LIBDE265_API __declspec(dllexport)
+#ifndef LIBDE265_STATIC_BUILD
+  #ifdef _WIN32
+    #ifdef LIBDE265_EXPORTS
+      #define LIBDE265_API __declspec(dllexport)
+    #else
+      #define LIBDE265_API __declspec(dllimport)
+    #endif
+  #elif HAVE_VISIBILITY
+    #define LIBDE265_API __attribute__((__visibility__("default")))
   #else
-  #define LIBDE265_API __declspec(dllimport)
-  #endif
-#elif HAVE_VISIBILITY
-  #ifdef LIBDE265_EXPORTS
-  #define LIBDE265_API __attribute__((__visibility__("default")))
-  #else
-  #define LIBDE265_API
+    #define LIBDE265_API
   #endif
 #else
   #define LIBDE265_API
